backend/controllers/taskController.js

```diff
@@ -15,4 +18,174 @@ const createTask = asyncHandler(async (req, res) => {
   }
 });
 
-module.exports = { createTask };
+const getTasks = asyncHandler(async (req, res) => {
+  const header = req.headers["authentication"];
+  const token = header.split(" ")[1];
+
+  const user_data = jwt.decode(token);
+
+
+  const userId = user_data._id;
+  const tasks = await Task.find({ user_id: userId }, (err, docs) => {
+    if(err) {
+      res.status(500);
+      throw new Error(`Could not fetch doc ${docs}`)
+    }
+  });
+
+  res.status(200).json(tasks);
+});
+
+const getTasksById = asyncHandler(async (req, res) => {
+  // Task id Parameter
+  const header = req.headers["authentication"];
+  const token = header.split(" ")[1];
+
+  const user_data = jwt.decode(token);
+
+
+  const taskId = req.params.id;
+
+  // Users id
+
+  const userId = user_data._id;
+
+  const tasks = await Task.find({
+    _id: taskId,
+    user_id: userId,
+  }, (err, docs) => {
+    if(err) {
+      res.status(500);
+      throw new Error(`Could not fetch doc ${docs}`)
+    }
+  });
+
+  res.status(200).json(tasks);
+});
+
+const getTasksByDay = asyncHandler(async (req, res) => {
+  const header = req.headers["authentication"];
+  const token = header.split(" ")[1];
+
+  const user_data = jwt.decode(token);
+
+  const userId = user_data._id;
+
+  const date = req.params.day; // get a day -> 26
+  // Users id -> Extract from JWT since GET does not take any body
+  let startDate = 0;
+  let endDate = 0;
+
+  if (date.length === 8) {
+    // parse
+    const year = parseInt(date.substring(0, 4));
+    const month = parseInt(date.substring(4, 6));
+    const day = parseInt(date.substring(6, 8));
+
+    try {
+      startDate = new Date(year, month - 1, day);
+      endDate = new Date(year, month - 1, day + 2);
+    } catch (error) {
+      res.status(400);
+      throw new Error("Invalid Date Input");
+
+    }
+  }
+
+  const tasks = await Task.find({
+    user_id: ObjectId(userId),
+    startDate: {
+      // $gte: startDate,
+      $gte: new Date(startDate),
+    },
+    endDate: {
+      $lte: new Date(endDate),
+    },
+    isStarted: true,
+  }, (err, docs) => {
+    if(err) {
+      res.status(500);
+      throw new Error(`Could not fetch doc ${docs}`)
+    }
+  });
+
+  res.status(200).json(tasks);
+});
+
+const startTask = asyncHandler(async (req, res) => {
```

Comment:
Since `startTask` and `endTask` are almost identical apart from toggling of the `isStarted` and updating either `taskStartedAt` and `taskEndedAt`, we can merge the two functions together.

We can check if the task is already started (through `isStarted`) and then this will determine whether we need to update `taskStartedAt` or `taskEndedAt`.

Reply:
One other thing to check is, what would happen that if we try to start a task after it was already ended. Does this mean that we change the start date? Or should that stay the same but we just reset the taskEndedAt time?

I think it should be that we should keep the `taskStartedAt` time and reset the `taskEndedAt` time.

Reply:
> One other thing to check is, what would happen that if we try to start a task after it was already ended. Does this mean that we change the start date? Or should that stay the same but we just reset the taskEndedAt time?
> 
> I think it should be that we should keep the `taskStartedAt` time and reset the `taskEndedAt` time.

A user should not be able to start a task that has already ended. I believe that the functionality to start an already ended task should be disabled in the frontend.

Reply:
Although it's partially true that the functionality to start an already ended task should be disabled in the frontend, it's also the backend's job to ensure this check as well.

Reply:
It's not a good practice to rely on the frontend for an edge case like this